common-ts/src/clients/marketDataFeed.ts

```diff
@@ -0,0 +1,710 @@
+	public readonly config: SubscriptionConfig;
+	protected _apiSubscription: ApiSubscription;
+	protected subject: Subject<T>;
+	private static idIncrementer: number = 0;
+
+	public get subscription(): ApiSubscription {
+		return this._apiSubscription;
+	}
+
+	private generateSubscriberId(): SubscriberId {
+		return `${this.config?.type}:${this.config?.marketSymbol}:${
+			this.config?.env.key
+		}:(${_Subscriber.idIncrementer++})` as SubscriberId;
+	}
+
+	constructor(config: SubscriptionConfig) {
+		this.config = { ...config }; // Make a copy of the config to avoid mutability issues
+		this.id = this.generateSubscriberId();
+	}
+
+	setSubscription(subscription: ApiSubscription) {
+		this._apiSubscription = subscription;
+	}
+
+	next(data: T) {
+		this.subject.next(data);
+	}
+}
+
+class CandleSubscriber extends _Subscriber<JsonCandle> {
+	public readonly config: CandleSubscriptionConfig;
+	public readonly subscriberSubscription: CandleSubscriberSubscription;
+
+	constructor(config: CandleSubscriptionConfig) {
+		super(config);
+		this.subject = new Subject<JsonCandle>();
+		this.subscriberSubscription = new CandleSubscriberSubscription(
+			this.id,
+			this.subject
+		);
+	}
+}
+
+class TradeSubscriber extends _Subscriber<JsonTrade[]> {
+	public readonly config: TradeSubscriptionConfig;
+	public readonly subscriberSubscription: TradeSubscriberSubscription;
+
+	constructor(config: TradeSubscriptionConfig) {
+		super(config);
+		this.subject = new Subject<JsonTrade[]>();
+		this.subscriberSubscription = new TradeSubscriberSubscription(
+			this.id,
+			this.subject
+		);
+	}
+}
+
+type Subscriber = CandleSubscriber | TradeSubscriber;
+
+function getCompatibleCandleSubscriptionLookupKey(
+	config: Pick<CandleSubscriptionConfig, 'marketSymbol' | 'resolution' | 'env'>
+): CandleSubscriptionLookupKey {
+	return `${config.marketSymbol}:${config.resolution}:${config.env.key}` as CandleSubscriptionLookupKey;
+}
+
+function getCompatibleTradeSubscriptionLookupKey(
+	config: Pick<CandleSubscriptionConfig, 'marketSymbol' | 'env'>
+): TradeSubscriptionLookupKey {
+	return `${config.marketSymbol}:${config.env.key}` as TradeSubscriptionLookupKey;
+}
+
+class ApiSubscription {
+	public readonly id: ApiSubscriptionId;
+	public readonly candleSubscribers: Map<SubscriberId, CandleSubscriber>;
+	public readonly tradeSubscribers: Map<SubscriberId, TradeSubscriber>;
+	public readonly config: Pick<
+		CandleSubscriptionConfig,
+		'marketSymbol' | 'resolution' | 'env'
+	>;
+	private apiClient: DataApiWsClient;
+	private onNoMoreSubscribers: () => void;
+	private static idIncrementer: number = 0;
+
+	private generateSubscriptionId(): ApiSubscriptionId {
+		return `${getCompatibleCandleSubscriptionLookupKey(
+			this.config
+		)}(${ApiSubscription.idIncrementer++})` as ApiSubscriptionId;
+	}
+
+	get tradeSubscriptionsLookupKey() {
+		return getCompatibleTradeSubscriptionLookupKey(this.config);
+	}
+
+	get candleSubscriptionsLookupKey() {
```

Comment:
good to document these as "Any candle subscription key that matches this key can use this ApiSubscription"

same for trade